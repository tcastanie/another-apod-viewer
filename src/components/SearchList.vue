<template>
  <div class="c-search-list">
    <article
      v-for="apod in listComputed"
      :key="apod.date"
      class="c-search-list__item"
      @click="openApod(apod)"
    >
      <div
        class="c-search-list__item-preview"
        :style="{ backgroundImage: `url(${apod.img_url})` }"
        :alt="apod.title"
      ></div>
      <h5>{{ apod.title }}</h5>
      <small>{{ apod.date }}</small>
    </article>
  </div>
</template>

<script>
import { useStore } from "@/stores/apod";
export default {
  props: {
    list: {
      type: Array,
      required: true,
    },
  },
  setup() {
    const store = useStore();
    return { store };
  },
  computed: {
    listComputed() {
      return this.list.map((apod) => {
        return {
          ...apod,
          img_url: JSON.stringify(apod.thumbnail_url ?? apod.url),
        };
      });
    },
  },
  methods: {
    openApod(apod) {
      this.store.saveApod(apod);
      this.$router.push({ name: "date", params: { date: apod.date } });
    },
  },
};
</script>

<style lang="scss" scoped>
.c-search-list {
  display: grid;
  grid-gap: 20px;
  grid-template-columns: repeat(2, 1fr);
  grid-template-rows: repeat(2, 1fr);

  @media (max-width: 992px) {
    grid-template-columns: 1fr;
    grid-template-rows: repeat(4, 1fr);
  }

  &__item {
    display: grid;
    margin: 0;
    cursor: pointer;
    grid-column-gap: 10px;
    grid-template-columns: repeat(2, 1fr);
    grid-template-rows: repeat(2, 1fr);
    transition: all 0.2s ease-in-out;

    &:hover,
    &:active {
      background-color: var(--secondary);
      color: var(--secondary-inverse);
      transform: scale(101%);
    }
  }

  &__item-preview {
    background-position: center;
    background-repeat: no-repeat;
    background-size: contain;
    grid-area: 1 / 1 / 3 / 2;
  }
}
</style>
