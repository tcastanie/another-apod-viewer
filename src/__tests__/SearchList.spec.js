import { describe, it, expect, beforeEach } from "vitest";
import { mount, flushPromises } from "@vue/test-utils";
import { createRouter, createWebHistory } from "vue-router";
import { routes } from "@/router/index.js";
import SearchList from "@/components/SearchList.vue";
import { createTestingPinia } from "@pinia/testing";
import { testList } from "./constantsMockups";

let router;
beforeEach(async () => {
  router = createRouter({
    history: createWebHistory(),
    routes,
  });
});

describe("SearchList", () => {
  it("renders properly", () => {
    const wrapper = mount(SearchList, {
      props: {
        list: testList,
      },
      global: {
        plugins: [createTestingPinia()],
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("open picture correctly", async () => {
    router.push("/search");
    await router.isReady();
    const wrapper = mount(SearchList, {
      props: {
        list: testList,
      },
      global: {
        plugins: [router, createTestingPinia()],
      },
    });
    wrapper.findAll("article")[0].trigger("click");
    await flushPromises();
    expect(wrapper.vm.$route.path).toBe(`/${testList[0].date}`);
  });
});
