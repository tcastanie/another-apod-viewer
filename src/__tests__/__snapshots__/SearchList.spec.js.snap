// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchList > renders properly 1`] = `
"<div data-v-cda856e3=\\"\\" class=\\"c-search-list\\">
  <article data-v-cda856e3=\\"\\" class=\\"c-search-list__item\\">
    <div data-v-cda856e3=\\"\\" class=\\"c-search-list__item-preview\\" style=\\"background-image: url(https://apod.nasa.gov/apod/image/2207/AS11JK44-6633-34_1100px.jpg);\\" alt=\\"The Eagle Rises\\"></div>
    <h5 data-v-cda856e3=\\"\\">The Eagle Rises</h5><small data-v-cda856e3=\\"\\">2022-07-30</small>
  </article>
  <article data-v-cda856e3=\\"\\" class=\\"c-search-list__item\\">
    <div data-v-cda856e3=\\"\\" class=\\"c-search-list__item-preview\\" style=\\"background-image: url(https://apod.nasa.gov/apod/image/2207/M94_Hubble_960.jpg);\\" alt=\\"Starburst Galaxy M94 from Hubble\\"></div>
    <h5 data-v-cda856e3=\\"\\">Starburst Galaxy M94 from Hubble</h5><small data-v-cda856e3=\\"\\">2022-07-31</small>
  </article>
</div>"
`;
